protocols/jvss/jvss.go

```diff
@@ -15,645 +21,249 @@ import (
-	// The longterm shared private public key pair used in this JVSS.
-	// The idea is that you can keep this protocol instance as a longterm JVSS
-	// using it to distributively sign anything as long as it runs.
-	secret *poly.SharedSecret
-	// The schnorr struct used to sign / verify using the longterm key
-	schnorr *poly.Schnorr
-	// Threshold related to how the shares are generated and reconstructed
-	info poly.Threshold
-	// suite used
-	suite abstract.Suite
-	// key from this node
-	key config.KeyPair
-	// longterm-receiver of the deals
-	receiver *poly.Receiver
-	// doneChan
-	doneChan chan bool
-	// done flag
-	done bool
-	// done lock
-	doneLock *sync.Mutex
-	// how many deals are ok
-	goodDeal int
-}
+// Identifiers for long- and short-term shared secrets.
+const (
+	LTSS = "LTSS"
+	STSS = "STSS"
+)
 
-// NewJVSSProtocol returns a JVSSProtocol with the fields set. You can then
-// change the fields or set specific ones etc. If you want to use JVSSProotocol
-// directly with SDA, you just need to  register this function:
-// ```func(h,t,tok) ProtocolInstance  { return NewJVSSProtocol(h,t,tok) }```
-// For example, this function returns a JVSSProtocol with a default
-// poly.Threshold. You can give a new one after calling this function.
-func NewJVSSProtocol(n *sda.Node) (*JVSSProtocol, error) {
-	// find ourselves in the entityList
-	var idx int = -1
-	// at the same time create the public list
-	tree := n.Tree()
-	nodes := tree.List()
-	pubs := make([]abstract.Point, len(nodes))
+// JVSS is the main protocol struct and implements the sda.ProtocolInstance
+// interface.
+type JVSS struct {
+	*sda.Node                         // The SDA TreeNode
+	keyPair     *config.KeyPair       // KeyPair of the host
+	nodeList    []*sda.TreeNode       // List of TreeNodes in the JVSS group
+	pubKeys     []abstract.Point      // List of public keys of the above TreeNodes
+	info        poly.Threshold        // JVSS thresholds
+	schnorr     *poly.Schnorr         // Long-term Schnorr struct to compute distributed signatures
+	secrets     map[string]*Secret    // Shared secrets (long- and short-term ones)
+	ltssInit    bool                  // Indicator whether shared secret has been already initialised or not
+	secretsDone chan bool             // Channel to indicate when shared secrets of all peers are ready
+	sigChan     chan *poly.SchnorrSig // Channel for JVSS signature
+}
+
+// Secret contains all information for long- and short-term shared secrets.
+type Secret struct {
+	secret   *poly.SharedSecret              // Shared secret
+	receiver *poly.Receiver                  // Receiver to aggregate deals
+	deals    map[int]*poly.Deal              // Buffer for deals
+	sigs     map[int]*poly.SchnorrPartialSig // Buffer for partial signatures
+	numConfs int                             // Number of collected confirmations that shared secrets are ready
+	mtx      *sync.Mutex                     // Mutex to sync access to numConfs
```

Comment:
Mutex are exceptions with regard to the *use pointers* rule. If you define them as values, you don't even have to initialize them (like in line 167)